src/Forms/GridField/GridFieldDetailForm_ItemRequest.php

```diff
@@ -410,6 +467,93 @@ public function doSave($data, $form)
         return $this->redirectAfterSave($isNewRecord);
     }
 
+    /**
+     * Goes to the previous record
+     * @param  array $data The form data
+     * @param  Form $form The Form object
+     * @return HTTPResponse
+     */
+    public function doPrevious($data, $form)
+    {
+        Controller::curr()->getResponse()->addHeader("X-Pjax", "Content");
+        $link = $this->getEditLink($this->getPreviousRecordID());
+        return Controller::curr()->redirect($link);
+    }
+
+    /**
+     * Goes to the next record
+     * @param  array $data The form data
+     * @param  Form $form The Form object
+     * @return HTTPResponse
+     */
+    public function doNext($data, $form)
+    {
+        Controller::curr()->getResponse()->addHeader("X-Pjax", "Content");
+        $link = $this->getEditLink($this->getNextRecordID());
+        return Controller::curr()->redirect($link);
+    }
+
+    /**
+     * Creates a new record. If you're already creating a new record,
+     * this forces the URL to change.
+     *
+     * @param  array $data The form data
+     * @param  Form $form The Form object
+     * @return HTTPResponse
+     */
+    public function doNew($data, $form)
+    {
+        return Controller::curr()->redirect(Controller::join_links($this->gridField->Link("item"), "new"));
+    }
+
+    /**
+     * Gets the edit link for a record
+     *
+     * @param  int $id The ID of the record in the GridField
+     * @return string
+     */
+    public function getEditLink($id)
+    {
+        return Controller::join_links(
+            $this->gridField->Link(),
+            "item",
+            $id,
+            '?gridState=' . urlencode($this->gridField->getState(false)->Value())
+        );
+    }
+
+    /**
+     * Gets the ID of the previous record in the list.
+     *
+     * @return int
+     */
+    public function getPreviousRecordID()
+    {
+        $gridField = $this->gridField;
+        $gridStateStr = $this->getRequest()->postVar('gridState');
+        $gridField->getState(false)->setValue($gridStateStr);
+
+        $map = $gridField->getManipulatedList()->limit(PHP_INT_MAX, 0)->column('ID');
+        $offset = array_search($this->record->ID, $map);
+        return isset($map[$offset-1]) ? $map[$offset-1] : false;
```

Comment:
OK, this amount of computation, while good enough for my module, isn't good enough for core. If the grid has 5,000 records in it, this falls on its face. Maybe even at 500 or 1,000.

Not sure what to do here, but here's a thought

* Add `itemsPerPage` to the `GridFieldPaginator` state (currently only stores `currentPage`)
* Instead of using an unlimited query on `getManipulatedList`, use `$itemsPerPage + 2` for the limit (to allow for the last record on the previous page and the first record on the next page), and use `($itemsPerPage * $currentPage) - 1` as the offset. In other words, with 10 records per page, on page 5, it should be looking at records 49-61.
* Search for the ID per the next couple lines, as is currently shown.

What do you think?